File: percommand.lua
local ____lualib = require("lualib_bundle")
local __TS__StringStartsWith = ____lualib.__TS__StringStartsWith
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local __TS__ObjectEntries = ____lualib.__TS__ObjectEntries
local __TS__ArrayUnshift = ____lualib.__TS__ArrayUnshift
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__StringTrim = ____lualib.__TS__StringTrim
local __TS__ArrayPushArray = ____lualib.__TS__ArrayPushArray
local ____exports = {}
local ____utils = require("utils")
local utils = ____utils.default
local BanType = BanType or ({})
BanType.OFFLINE_ONLY = "offline-only"
BanType.ONLINE_ONLY = "online-only"
BanType.ARGUMENTS = "arguments"
BanType.COOLDOWN = "cooldown"
BanType.BLACKLIST = "blacklist"
BanType.REMINDER_PREVENTION = "reminder-prevention"
local commands = {}
do
    --- COMMON = [user:] [command:] [invocation:]
    -- $ban [type:blacklist] $COMMON
    -- $ban type:reminder-prevention [user:] [channel:]
    -- $ban type:cooldown [multiplier:] $COMMON
    -- $ban type:arguments [all:(true|false)] [index:] [string:]
    -- $unban type:arguments [clear:(true|false)] [all:(true|false)] [index:] [string:]
    function commands.ban(cmd, invocation, text, prefix)
        local out = utils.new_completion_list()
        if __TS__StringStartsWith(prefix, "command:") then
            return utils.commands_and_their_aliases("command:")
        end
        if __TS__StringStartsWith(prefix, "invocation:") then
            return utils.commands_and_their_aliases("invocation:")
        end
        local ____utils_parse_params_3 = utils.parse_params
        local ____text_2 = text
        local ____opt_0 = cmd.params
        local parsed = ____utils_parse_params_3(
            ____text_2,
            ____opt_0 and __TS__ArrayMap(
                cmd.params,
                function(____, it) return it.name end
            ) or ({})
        )
        local typ = parsed.params:get("type")
        local invalid_type = true
        for ____, ____value in ipairs(__TS__ObjectEntries(BanType)) do
            local k = ____value[1]
            local v = ____value[2]
            if v == typ then
                invalid_type = false
            end
        end
        if typ == nil or invalid_type then
            local ____out_values_4 = out.values
            ____out_values_4[#____out_values_4 + 1] = "type:" .. BanType.ARGUMENTS
            local ____out_values_5 = out.values
            ____out_values_5[#____out_values_5 + 1] = "type:" .. BanType.BLACKLIST
            local ____out_values_6 = out.values
            ____out_values_6[#____out_values_6 + 1] = "type:" .. BanType.COOLDOWN
            local ____out_values_7 = out.values
            ____out_values_7[#____out_values_7 + 1] = "type:" .. BanType.OFFLINE_ONLY
            local ____out_values_8 = out.values
            ____out_values_8[#____out_values_8 + 1] = "type:" .. BanType.ONLINE_ONLY
            local ____out_values_9 = out.values
            ____out_values_9[#____out_values_9 + 1] = "type:" .. BanType.REMINDER_PREVENTION
            typ = "blacklist"
        end
        if typ == BanType.REMINDER_PREVENTION then
            local ____out_values_10 = out.values
            ____out_values_10[#____out_values_10 + 1] = "user:"
            local ____out_values_11 = out.values
            ____out_values_11[#____out_values_11 + 1] = "channel:"
            out.hide_others = false
            return out
        end
        if typ == BanType.COOLDOWN then
            local ____out_values_12 = out.values
            ____out_values_12[#____out_values_12 + 1] = "multiplier:"
            local ____out_values_13 = out.values
            ____out_values_13[#____out_values_13 + 1] = "multiplier:1.5"
            local ____out_values_14 = out.values
            ____out_values_14[#____out_values_14 + 1] = "multiplier:2"
        end
        if typ == BanType.ARGUMENTS then
            local ____out_values_15 = out.values
            ____out_values_15[#____out_values_15 + 1] = "all:true"
            local ____out_values_16 = out.values
            ____out_values_16[#____out_values_16 + 1] = "index:"
            local ____out_values_17 = out.values
            ____out_values_17[#____out_values_17 + 1] = "string:"
            if invocation == "unban" then
                local ____out_values_18 = out.values
                ____out_values_18[#____out_values_18 + 1] = "clear:true"
            end
        end
        local ____out_values_19 = out.values
        ____out_values_19[#____out_values_19 + 1] = "user:"
        local ____out_values_20 = out.values
        ____out_values_20[#____out_values_20 + 1] = "command:"
        local ____out_values_21 = out.values
        ____out_values_21[#____out_values_21 + 1] = "invocation:"
        out.hide_others = true
        return out
    end
    --- Advanced mode: $block command:<all|blockable command> user: [channel:] [platform:]
    -- Simple mode: $block <user> <all|blockable command>
    function commands.block(cmd, text, prefix)
        local ____utils_parse_params_25 = utils.parse_params
        local ____text_24 = text
        local ____opt_22 = cmd.params
        local parsed = ____utils_parse_params_25(
            ____text_24,
            ____opt_22 and __TS__ArrayMap(
                cmd.params,
                function(____, it) return it.name end
            ) or ({})
        )
        local argv = parsed.argv
        table.remove(argv, 1)
        if __TS__StringStartsWith(prefix, "command:") then
            return utils.commands_and_their_aliases("command:", {"BLOCK"})
        end
        local out = utils.new_completion_list()
        local ____out_values_26 = out.values
        ____out_values_26[#____out_values_26 + 1] = "channel:"
        local ____out_values_27 = out.values
        ____out_values_27[#____out_values_27 + 1] = "command:"
        local ____out_values_28 = out.values
        ____out_values_28[#____out_values_28 + 1] = "platform:"
        local ____out_values_29 = out.values
        ____out_values_29[#____out_values_29 + 1] = "user:"
        if parsed.params.size == 0 then
            local merged = utils.commands_and_their_aliases("", {"BLOCK"})
            __TS__ArrayUnshift(merged.values, "all")
            merged.hide_others = #argv == 2
            return merged
        end
        return utils.new_completion_list()
    end
    --- Advanced mode: $unping command:<all|command> [user:] [channel:] [platform:]
    -- Simple mode: $unping <all|command>
    function commands.unping(cmd, text, prefix)
        if __TS__StringStartsWith(prefix, "command:") then
            return utils.commands_and_their_aliases("command:")
        end
        local parsed = utils.parse_params(text, {"channel", "user", "command", "platform"})
        local argv = parsed.argv
        table.remove(argv, 1)
        local out = utils.new_completion_list()
        local ____out_values_30 = out.values
        ____out_values_30[#____out_values_30 + 1] = "channel:"
        local ____out_values_31 = out.values
        ____out_values_31[#____out_values_31 + 1] = "user:"
        local ____out_values_32 = out.values
        ____out_values_32[#____out_values_32 + 1] = "command:"
        local ____out_values_33 = out.values
        ____out_values_33[#____out_values_33 + 1] = "platform:"
        if parsed.params.size == 0 then
            local merged = utils.commands_and_their_aliases("")
            __TS__ArrayUnshift(merged.values, "all")
            __TS__ArrayUnshift(
                merged.values,
                table.unpack(out.values)
            )
            merged.hide_others = true
            return merged
        end
        return out
    end
    --- $optout id:<number>
    -- $optout [channel:] [platform:] [command:(all|optoutable command)]
    -- $optout <all|optoutable command> [channel:] [platform:]
    function commands.optout(cmd, text, prefix, cursor_position)
        if __TS__StringStartsWith(prefix, "command:") then
            local cmds = utils.commands_and_their_aliases("command:", {"OPT_OUT"})
            __TS__ArrayUnshift(cmds.values, "command:all")
            return cmds
        end
        local ____opt_34 = cmd.params
        local params = ____opt_34 and __TS__ArrayMap(
            cmd.params,
            function(____, it) return it.name end
        ) or ({})
        local parsed = utils.parse_params(text, params)
        local argv = parsed.argv
        local committed_args = __TS__StringTrim(__TS__StringSubstring(text, 0, cursor_position - #prefix) .. __TS__StringSubstring(text, cursor_position))
        local committed = utils.parse_params(committed_args, params)
        table.remove(committed.argv, 1)
        table.remove(argv, 1)
        local out = utils.new_completion_list()
        if parsed.params:get("id") == nil then
            local ____out_values_36 = out.values
            ____out_values_36[#____out_values_36 + 1] = "id:"
            local ____out_values_37 = out.values
            ____out_values_37[#____out_values_37 + 1] = "channel:"
            local ____out_values_38 = out.values
            ____out_values_38[#____out_values_38 + 1] = "platform:"
            if #committed.argv == 0 then
                local ____out_values_39 = out.values
                ____out_values_39[#____out_values_39 + 1] = "command:"
                local to_merge = utils.commands_and_their_aliases("")
                __TS__ArrayPushArray(out.values, to_merge.values)
            end
        end
        return out
    end
    function commands.unmention(cmd, text, prefix)
        if __TS__StringStartsWith(prefix, "command:") then
            local cmds = utils.commands_and_their_aliases("command:")
            return cmds
        end
        local ____utils_parse_params_43 = utils.parse_params
        local ____text_42 = text
        local ____opt_40 = cmd.params
        local parsed = ____utils_parse_params_43(
            ____text_42,
            ____opt_40 and __TS__ArrayMap(
                cmd.params,
                function(____, it) return it.name end
            ) or ({})
        )
        local out = utils.new_completion_list()
        out.hide_others = true
        local ____out_values_44 = out.values
        ____out_values_44[#____out_values_44 + 1] = "channel:"
        local ____out_values_45 = out.values
        ____out_values_45[#____out_values_45 + 1] = "command:"
        local ____out_values_46 = out.values
        ____out_values_46[#____out_values_46 + 1] = "platform:"
        if parsed.params.size == 0 then
            local to_merge = utils.commands_and_their_aliases("")
            out.hide_others = true
            __TS__ArrayPushArray(out.values, to_merge.values)
        end
        return out
    end
end
____exports.default = commands
return ____exports

File: percommand.ts
import {Command} from "./init";
import utils from "./utils";
enum BanType {
    OFFLINE_ONLY = "offline-only",
    ONLINE_ONLY = "online-only",
    ARGUMENTS = "arguments",
    COOLDOWN = "cooldown",
    BLACKLIST = "blacklist",
    REMINDER_PREVENTION = "reminder-prevention",
};

namespace commands {
    /**
     * COMMON = [user:] [command:] [invocation:]
     * $ban [type:blacklist] $COMMON
     * $ban type:reminder-prevention [user:] [channel:]
     * $ban type:cooldown [multiplier:] $COMMON
     * $ban type:arguments [all:(true|false)] [index:] [string:]
     * $unban type:arguments [clear:(true|false)] [all:(true|false)] [index:] [string:]
     */
    export function ban(this: void, cmd: Command, invocation: string, text: string, prefix: string): c2.CompletionList {
        const out = utils.new_completion_list();
        if (prefix.startsWith("command:")) {
            return utils.commands_and_their_aliases("command:");
        }
        if (prefix.startsWith("invocation:")) {
            return utils.commands_and_their_aliases("invocation:");
        }

        const parsed = utils.parse_params(text, cmd.params?.map(it => it.name) ?? []);
        let typ = parsed.params.get("type");
        let invalid_type = true;
        for (const [k, v] of Object.entries(BanType)) {
            if (v === typ) {
                invalid_type = false;
            }
        }
        if (typ === undefined || invalid_type) {
            // User didn't specify a "type:" param, suggest them possible variants
            // Not having one is perfectly valid too.
            // Note that lua k-v tables don't guarantee order
            out.values.push("type:" + BanType.ARGUMENTS);
            out.values.push("type:" + BanType.BLACKLIST);
            out.values.push("type:" + BanType.COOLDOWN);
            out.values.push("type:" + BanType.OFFLINE_ONLY);
            out.values.push("type:" + BanType.ONLINE_ONLY);
            out.values.push("type:" + BanType.REMINDER_PREVENTION);

            typ = "blacklist";
        }

        if (typ === BanType.REMINDER_PREVENTION) {
            out.values.push("user:");
            out.values.push("channel:");
            out.hide_others = false;
            return out;
        }
        if (typ === BanType.COOLDOWN) {
            out.values.push("multiplier:");
            out.values.push("multiplier:1.5");
            out.values.push("multiplier:2");
        }
        if (typ === BanType.ARGUMENTS) {
            out.values.push("all:true");
            out.values.push("index:");
            out.values.push("string:");
            if (invocation === 'unban') {
                out.values.push("clear:true");
            }
        }

        out.values.push("user:");
        out.values.push("command:");
        out.values.push("invocation:");
        out.hide_others = true;
        return out;
    }

    /**
     * Advanced mode: $block command:<all|blockable command> user: [channel:] [platform:]
     * Simple mode: $block <user> <all|blockable command>
     */
    export function block(this: void, cmd: Command, text: string, prefix: string): c2.CompletionList {
        const parsed = utils.parse_params(text, cmd.params?.map(it => it.name) ?? []);
        const {argv} = parsed;
        argv.shift(); // remove command name

        // maybe simple mode?
        if (prefix.startsWith("command:")) {
            return utils.commands_and_their_aliases("command:", ["BLOCK"]);
        }
        const out = utils.new_completion_list();
        out.values.push("channel:");
        out.values.push("command:");
        out.values.push("platform:");
        out.values.push("user:");
        if (parsed.params.size === 0) {
            // we can be in simple mode, as we have no params
            const merged = utils.commands_and_their_aliases("", ["BLOCK"]);
            merged.values.unshift("all");
            merged.values.push(...out.values);
            // have $[un]block <username> <word>|
            merged.hide_others = argv.length === 2;
            return merged;
        }
        return out;
    }

    /**
     * Advanced mode: $unping command:<all|command> [user:] [channel:] [platform:]
     * Simple mode: $unping <all|command>
     */
    export function unping(this: void, cmd: Command, text: string, prefix: string): c2.CompletionList {
        if (prefix.startsWith("command:")) {
            return utils.commands_and_their_aliases("command:");
        }
        // XXX: This is parsed in-command
        const parsed = utils.parse_params(text, ["channel", "user", "command", "platform"]);
        const {argv} = parsed;
        argv.shift(); // remove command name

        const out = utils.new_completion_list();
        out.values.push("channel:");
        out.values.push("user:");
        out.values.push("command:");
        out.values.push("platform:");
        if (parsed.params.size === 0) {
            // simple mode
            // $unping <all|command>
            const merged = utils.commands_and_their_aliases("");
            merged.values.unshift("all");
            // allow user to pick advanced mode but not complete usernames, emotes and such
            merged.values.unshift(...out.values);
            merged.hide_others = true;
            return merged;
        }
        return out;
    }

    /**
     * $optout id:<number>
     * $optout [channel:] [platform:] [command:(all|optoutable command)]
     * $optout <all|optoutable command> [channel:] [platform:]
     */
    export function optout(this: void, cmd: Command, text: string, prefix: string, cursor_position: number): c2.CompletionList {
        if (prefix.startsWith("command:")) {
            const cmds = utils.commands_and_their_aliases("command:", ["OPT_OUT"]);
            cmds.values.unshift("command:all");
            return cmds;
        }
        const params = cmd.params?.map(it => it.name) ?? [];
        const parsed = utils.parse_params(text, params);
        const {argv} = parsed;
        const committed_args = (text.substring(0, cursor_position - prefix.length) + text.substring(cursor_position)).trim();
        const committed = utils.parse_params(committed_args, params);
        committed.argv.shift();
        argv.shift(); // remove command name
        const out = utils.new_completion_list();
        // ID logically conflicts with all other params
        if (parsed.params.get("id") === null) {
            out.values.push("id:");
            out.values.push('channel:');
            out.values.push('platform:');
            // no command
            if (committed.argv.length === 0) {
                out.values.push('command:');
                const to_merge = utils.commands_and_their_aliases("");
                out.values.push(...to_merge.values);
            }
        }
        return out;
    }
    export function unmention(this: void, cmd: Command, text: string, prefix: string): c2.CompletionList {
        if (prefix.startsWith("command:")) {
            const cmds = utils.commands_and_their_aliases("command:");
            //cmds.values.unshift("command:all");
            return cmds;
        }
        const parsed = utils.parse_params(text, cmd.params?.map(it => it.name) ?? []);
        const out = utils.new_completion_list();
        out.hide_others = true;
        out.values.push("channel:");
        out.values.push("command:");
        out.values.push("platform:");

        if (parsed.params.size === 0) {
            const to_merge = utils.commands_and_their_aliases("");
            out.hide_others = true;
            out.values.push(...to_merge.values);
        }
        return out;
    }

    const PLAYER_COMMON = [
        'rude:',
        'seasonal:',
    ];
    /**
      * $osrs itemid <item>
      * $osrs price <item>
      *
      * PLAYER_COMMON := [rude:] [seasonal:]
      * $osrs <username> [skill:] [virtual:] ${PLAYER_COMMON}
      * ACTIVITY := ?
      * $osrs (kc|killcount|kill-count) (activity:${ACTIVITY}|boss:${ACTIVITY}) ${PLAYER_COMMON} <name> 
      * $osrs stars|star
      * $osrs status
      * $osrs guthix|tears|tog
      * $osrs wiki <free form text>
      */

    export function osrs(this: void, cmd: Command, text: string, prefix: string): c2.CompletionList {
        const s = text.split(" ", 2);
        const sub = s[1];

        const out = utils.new_completion_list();
        if (sub === 'itemid' || sub === 'price'
            || sub === 'stars' || sub === 'star'
            || sub === 'status'
            || sub === 'guthix' || sub == 'tears' || sub == 'tog'
            || sub == 'wiki') {
            return out; // nothing to complete
        }
        if (sub === 'kc' || sub == 'killcount' || sub === 'kill-count') {
            // TODO

            if (!text.includes('activity:') && !text.includes('boss:')) {
                out.values.push('activity:');
                out.values.push('boss:');
            }
            out.values.push(...PLAYER_COMMON);
            return out;
        }
        if (utils.count_occurences_of_byte(text, ' ') <= 2) {
            out.values.push('itemid ');
            out.values.push('price ');
            out.values.push('stars ', 'star ');
            out.values.push('status ');
            out.values.push('guthix ', 'tears ', 'tog ');
            out.values.push('wiki ');
            out.values.push('kc ', 'killcount ', 'kill-count ');
            out.values.push(...PLAYER_COMMON);
            out.values.push('skill:', 'virtual:');
            return out;
        }
        mm2pl_open_debugger().break_here();
        // $osrs <sub> ...
        return out;
    }
}

export default commands;
